### ISSUE TITLE
Expression tree parsing fails after solution migration from MVC4 (framework 4.6.1)

### ISSUE THREAD

**Issue by rynowak** [aspnet/Mvc#6634]
_From @wurdum on August 8, 2017 14:53_

After migration to asp net core previously working expression tree parsing fails now. 

I use expressions to generate qualified names for inputs. This code `@Model.GetElementName(m => m.Rows.Get(row).Cells.Get(cell).Value)` should generate something like `Rows[1].Cells[1].Value`

What wonders me, this code works as expected:

```csharp
@foreach (var r in Model.Rows) {
    foreach (var c in r.Cells) {
        var row = r;
        var cell = c;
        <div>
            <p>Row: @row.Id, Cell: @cell.Id</p>
            <p>Input name: @Model.GetElementName(m => m.Rows.Get(row).Cells.Get(cell).Value)</p>
        </div>
    }
}
```

But this fails:

```csharp
@foreach (var row in Model.Rows) {
    foreach (var cell in row.Cells) {
        <div>
            <p>Row: @row.Id, Cell: @cell.Id</p>
            <p>Input name: @Model.GetElementName(m => m.Rows.Get(row).Cells.Get(cell).Value)</p>
        </div>
    }
}
```

I've created repo with project that reproduces error: [https://github.com/wurdum/ExpressionTreeIssue](https://github.com/wurdum/ExpressionTreeIssue)

_Copied from original issue: aspnet/Razor#1603_

**Comment by dougbu** [comment]
I don't believe there's a framework bug here. Certainly not a new MVC issue. Looks more like a minor change in the generated code for some complex Linq expressions.

The root problem is an `ArgumentException` thrown in the repro project's `Parser` class while it is visiting nodes in a Linq expression. The exception occurs after `"Value"` and `"Cells[1]"` are already in `steps` i.e. when visiting `the `m.Rows.Get(rows)` portion.

Put another way, the `Parser` should evaluate sub-`Expression`s more carefully. See for example the way [`ExpressionHelper`](https://github.com/aspnet/Mvc/blob/5fee1cac83d5e41a3027a7f30d87764e18850c54/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionHelper.cs) isolates indexer expressions before evaluating them. Of course `ExpressionHelper` does not handle arbitrary method calls.

```
System.ArgumentException: Field 'row' defined on type 'AspNetCore._Views_Home_Index_cshtml+<>c__DisplayClass21_0' is not a field on the target object which is of type 'AspNetCore._Views_Home_Index_cshtml+<>c__DisplayClass21_1'.
   at System.Reflection.RtFieldInfo.CheckConsistency(Object target)
   at System.Reflection.RtFieldInfo.InternalGetValue(Object obj, StackCrawlMark& stackMark)
   at System.Reflection.RtFieldInfo.GetValue(Object obj)
   at ExpressionTreeIssue.Controllers.Parser`1.GetArgumentId(MemberExpression arg) in C:\dd\Projects\ExpressionTreeIssue\Controllers\Parser.cs:line 63
   at ExpressionTreeIssue.Controllers.Parser`1.VisitMethod(MemberExpression caller, MemberExpression arg, List`1 segments) in C:\dd\Projects\ExpressionTreeIssue\Controllers\Parser.cs:line 58
   at ExpressionTreeIssue.Controllers.Parser`1.VisitNode(Expression body, List`1 steps) in C:\dd\Projects\ExpressionTreeIssue\Controllers\Parser.cs:line 44
   at ExpressionTreeIssue.Controllers.Parser`1.VisitNode(Expression body, List`1 steps) in C:\dd\Projects\ExpressionTreeIssue\Controllers\Parser.cs:line 47
   at ExpressionTreeIssue.Controllers.Parser`1.VisitNode(Expression body, List`1 steps) in C:\dd\Projects\ExpressionTreeIssue\Controllers\Parser.cs:line 36
   at ExpressionTreeIssue.Controllers.Parser`1.GetCallSteps() in C:\dd\Projects\ExpressionTreeIssue\Controllers\Parser.cs:line 19
   at ExpressionTreeIssue.Controllers.Extensions.GetElementName[T](T model, Expression`1 expression) in C:\dd\Projects\ExpressionTreeIssue\Controllers\Extensions.cs:line 10
   at AspNetCore._Views_Home_Index_cshtml.<ExecuteAsync>d__21.MoveNext() in /Views/Home/Index.cshtml:line 8
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.<RenderPageAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.<RenderAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
  at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.<ExecuteAsync>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.ViewResult.<ExecuteResultAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeResultAsync>d__30.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeNextResultFilterAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeNextResourceFilter>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Hosting.Internal.RequestServicesContainerMiddleware.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.Frame`1.<RequestProcessingAsync>d__2.MoveNext()
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
```

(I removed stack frames matching `System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()` and `System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)`.)

----

FYI seemingly-similar code using HTML helpers such as the following works as expected:
```
@foreach (var row in Model.Rows) {
    foreach (var cell in row.Cells) {
        <div>
            <p>Row: @row.Id, Cell: @cell.Id</p>
            <p>Input name: @Html.IdFor(m => cell.Value)</p>
        </div>
    }
}
```
or
```
@for (var row = 0; row < Model.Rows.Count; row++) {
    for (var cell = 0; cell < Model.Rows[row].Cells.Count; cell++) {
        <div>
            <p>Row: @Model.Rows[row].Id, Cell: @Model.Rows[row].Cells[cell].Id</p>
            <p>Input name: @Html.IdFor(m => m.Rows[row].Cells[cell].Value)</p>
        </div>
    }
}
```

**Comment by wurdum** [comment]
Thank you for reference to good example.
I've fixed Parser error ([e661137](https://github.com/wurdum/ExpressionTreeIssue/commit/e6611378f64bbaab9634747302f8b0485ebdd094)) using [ExpressionHelper](https://github.com/aspnet/Mvc/blob/5fee1cac83d5e41a3027a7f30d87764e18850c54/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionHelper.cs#L239)'s way to parse indexers.

**Comment by dougbu** [comment]
Closing this since all questions seem to have been answered. @wurdum please reopen or otherwise let us know if I'm missing something.
